programs/conditional_vault/src/state/conditional_vault.rs

```diff
@@ -9,27 +9,60 @@ pub enum VaultStatus {
 
 #[account]
 pub struct ConditionalVault {
-    pub status: VaultStatus,
-    /// The account that can either finalize the vault to make conditional tokens
-    /// redeemable for underlying tokens or revert the vault to make deposit
-    /// slips redeemable for underlying tokens.
-    pub settlement_authority: Pubkey,
-    /// The mint of the tokens that are deposited into the vault.
+    pub question: Pubkey,
     pub underlying_token_mint: Pubkey,
-    /// The vault's storage account for deposited funds.
     pub underlying_token_account: Pubkey,
-    pub conditional_on_finalize_token_mint: Pubkey,
-    pub conditional_on_revert_token_mint: Pubkey,
+    pub conditional_token_mints: Vec<Pubkey>,
     pub pda_bump: u8,
     pub decimals: u8,
 }
 
+impl ConditionalVault {
+    /// Checks that the vault's assets are always greater than its potential
+    /// liabilities. Should be called anytime you mint or burn conditional
+    /// tokens.
+    ///
+    /// `conditional_token_supplies` should be in the same order as
+    /// `vault.conditional_token_mints`.
+    pub fn invariant(
+        &self,
+        question: &Question,
+        conditional_token_supplies: Vec<u64>,
+        vault_underlying_balance: u64,
+    ) -> Result<()> {
+        // if the question isn't resolved, the vault should have more underlying
+        // tokens than ANY conditional token mint's supply
+
+        // if the question is resolved, the vault should have more underlying
+        // tokens than the sum of the conditional token mint's supplies multiplied
+        // by their respective payouts
+
+        let max_possible_liability = if !question.is_resolved() {
+            // safe because conditional_token_supplies is non-empty
+            *conditional_token_supplies.iter().max().unwrap()
+        } else {
+            conditional_token_supplies
+                .iter()
+                .enumerate()
+                .map(|(i, supply)| {
+                    *supply * question.payout_numerators[i] as u64
```

Comment:
in redeem_tokens.rs, you cast to u128 for safe multiplication. I'd recommend doing it here as well

Reply:
yep, good catch!

Reply:
fixed: https://github.com/metaDAOproject/futarchy/pull/246/commits/963ee252542ce8bc8b825f7a537a0203edd5a60f